### PR TITLE
[E2E] Add basics of the E2E suite

### LINKED ISSUE #16: [Testing] E2E Testing Framework
## Objective

Migrate the previously built E2E Testing framework to this repo.

## Origin Document

The [pocket repo](https://github.com/pokt-network/pocket) implements an [E2E testing framework](https://github.com/pokt-network/pocket/tree/main/e2e) that needs to be migrated to the poktroll repo.

## Goals

- Enable E2E regression testing
- Enable automation that'll be used in load testing

## Deliverables

- [x] A single PR that adapts [this directory from the v1 repo](https://github.com/pokt-network/pocket/tree/main/e2e) to this repository
- [x] A single PR that introduces an E2E test for staking/unstaking
- [ ] A single PR that introduces an E2E test for a relay that verifies on-chain claims & proofs

## Non-goals / Non-deliverables

- Provide the tooling for every type of E2E Test that we may need
- Automated E2E tests in our CI/CD pipeline

## General deliverables

- [ ] **Comments**: Add/update TODOs and comments alongside the source code so it is easier to follow.
- [ ] **Testing**: Add new tests (unit and/or E2E) to the test suite.
- [ ] **Makefile**: Add new targets to the Makefile to make the new functionality easier to use.
- [ ] **Documentation**: Update architectural or development READMEs; use [mermaid](https://mermaid-js.github.io/mermaid/) diagrams where appropriate.

---

**Creator**: @Olshansk 
**Co-Owners**: @bryanchriswhite @red-0ne @h5law

### PR DESCRIPTION
**Pull request by h5law**
## Summary

This PR ports over the basics of the gherkin suite from the old v1 repo. It contains a "help" test and a "send" test to ensure this works with k8s localnet.

### Summary generated by Reviewpad on 14 Oct 23 11:55 UTC
This pull request includes several changes across multiple files:

1. In the file `node.go`, a new file has been added that provides functionalities related to interacting with a node using the `pocketd` binary. It includes the declaration of a package named `e2e`, import statements for required packages, variable definitions, struct definitions, interface declaration, and function implementations.

2. A new file named `help.feature` has been added in the `e2e/tests` directory. It defines a feature for the root namespace and includes a scenario for when a user needs help with the `pocketd` binary. The scenario includes steps for checking if the user has the binary installed, running the "help" command, verifying the expected output, and validating the error-free exit of the binary.

3. Changes in the file `init_test.go` involve importing necessary packages and libraries, defining a regular expression, declaring a struct, and implementing multiple test functions related to end-to-end testing and interacting with `pocketd` binary.

4. Another new file named `send.feature` has been added in the `e2e/tests` directory. It defines a feature called "Tx Namespace" with a scenario named "User can send uPOKT". This scenario tests if the user can successfully send 10000 uPOKT to another address using the `pocketd` binary.

5. The `go.mod` file has been modified to add new dependencies and update existing ones.

6. In the `Makefile` diff, modifications include updates to the `POCKET_NODE` variable assignment, addition of a new target `test_e2e` for running end-to-end tests, modification of the `go_test` target, and exclusion of the `Makefile` itself from a search pattern.

These changes add new files, functionalities, and dependencies to the codebase and make improvements to existing files.

## Issue

Fixes (in part) [16](https://github.com/pokt-network/poktroll/issues/16) 

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [x] **Run all unit tests**: `make test_all_unit`
- [x] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have performed a self-review of my own code
- [x] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

File: e2e/tests/init_test.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+//go:build e2e
+
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/cucumber/godog"
```
Comment (bryanchriswhite) [inline e2e/tests/init_test.go:10]: We should prefer Regen Network's `gocuke` instead of `godog`:

https://github.com/regen-network/gocuke

**Review by okdas — APPROVED** [review]
I like that we removed an additional abstraction of a container here, as the test will run on a container on DevNet anyway. Thank you!

File: e2e/tests/send.feature [hunk 2]
```diff
@@ -0,0 +1,8 @@
+Feature: Tx Namespace
+
+  Scenario: User can send uPOKT
+    Given the user has the pocketd binary installed
+    When the user sends 10000 uPOKT to another address
+    Then the user should be able to see standard output containing "txhash:"
+    And the user should be able to see standard output containing "code: 0"
+    And the pocketd binary should exit without error
```
Comment (Olshansk) [inline e2e/tests/send.feature:8]: Couple of changes I'd make to this:
1. Validate which account has 10000 uPOKT. I don't know who the user is
2. Explicitly state WHEN `account a1 sends 10000 to account a2`
3. Verify that the final balance of a1 is lower
4. Verify that the final balance of a2 is higher

Now we have a real regression test.

**Review by Olshansk — COMMENTED** [review]
Took a look at this after it was merged. LGTM overall but requested a couple of improvements.

**Pull request merged**